New service: `DeviceRegistry` + new resources: `azurerm_device_registry_asset` and `azurerm_device_registry_asset_endpoint_profile`

mryanlo opened the pull request:
<!--  All Submissions -->
This PR introduces support for the Device Registry service, comprising of:

### Resources
```
- azurerm_device_registry_asset
- azurerm_device_registry_asset_endpoint_profile
```

## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->
Swagger: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/deviceregistry/resource-manager/Microsoft.DeviceRegistry/stable/2024-11-01

Azure Doc: https://learn.microsoft.com/en-us/azure/iot-operations/discover-manage-assets/howto-manage-assets-remotely?tabs=cli

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

The acceptance tests for `azurerm_device_registry_asset` and `azurerm_device_registry_asset_endpoint_profile` require an arc-enabled kubernetes cluster with Azure IoT Operations' service extension installed on the cluster in order to pass. Creating this cluster in the terraform acceptance tests has been successful but it takes over 45 minutes to complete. In order for the tests to run correctly, the following environment variables need to be set:
```bash
# ID of the Azure subscription that the acceptance tests will run on
export ARM_SUBSCRIPTION_ID=<subscription ID> 

# The Client ID of the Service Principal that will run the acceptance tests.
export ARM_CLIENT_ID=<client ID>

# The password of the Service Principal that will run the acceptance tests.
export ARM_CLIENT_SECRET=<client secret> 

# The Object ID of the Custom Locations RP's Entra App, as mentioned in previous step.
export ARM_ENTRA_APP_OBJECT_ID=<object ID> 
```

If one of those variables isn't set, the tests will be skipped.

Here are the successful test results. I had to run each test one at a time because my subscription did not have enough quota to create multiple VMs. Please ensure that the Azure subscription can create multiple VMs at once. Here are the final results:

ASSET ENDPOINT PROFILE TESTS
===

```
make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAssetEndpointProfile_basic'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAssetEndpointProfile_basic -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAssetEndpointProfile_basic
=== PAUSE TestAccAssetEndpointProfile_basic
=== CONT  TestAccAssetEndpointProfile_basic
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327203216201991'
Connecting to the VM with IP Address: 52.175.235.1
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAssetEndpointProfile_basic (2542.60s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2561.008s

make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAssetEndpointProfile_complete_certificate'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAssetEndpointProfile_complete_certificate -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAssetEndpointProfile_complete_certificate
=== PAUSE TestAccAssetEndpointProfile_complete_certificate
=== CONT  TestAccAssetEndpointProfile_complete_certificate
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327211658376605'
Connecting to the VM with IP Address: 40.125.75.160
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAssetEndpointProfile_complete_certificate (2558.27s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2563.423s

make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAssetEndpointProfile_complete_usernamePassword'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAssetEndpointProfile_complete_usernamePassword -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAssetEndpointProfile_complete_usernamePassword
=== PAUSE TestAccAssetEndpointProfile_complete_usernamePassword
=== CONT  TestAccAssetEndpointProfile_complete_usernamePassword
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327220155347867'
Connecting to the VM with IP Address: 52.191.190.38
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAssetEndpointProfile_complete_usernamePassword (2849.30s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2852.310s

$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAssetEndpointProfile_complete_anonymous'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAssetEndpointProfile_complete_anonymous -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAssetEndpointProfile_complete_anonymous
=== PAUSE TestAccAssetEndpointProfile_complete_anonymous
=== CONT  TestAccAssetEndpointProfile_complete_anonymous
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327113428795954'
Connecting to the VM with IP Address: 52.148.176.114
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAssetEndpointProfile_complete_anonymous (2705.41s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2723.071s

$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAssetEndpointProfile_requiresImport'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAssetEndpointProfile_requiresImport -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAssetEndpointProfile_requiresImport
=== PAUSE TestAccAssetEndpointProfile_requiresImport
=== CONT  TestAccAssetEndpointProfile_requiresImport
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327101031664339'
Connecting to the VM with IP Address: 4.155.66.141
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAssetEndpointProfile_requiresImport (3411.25s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        3430.707s

$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAssetEndpointProfile_update'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAssetEndpointProfile_update -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAssetEndpointProfile_update
=== PAUSE TestAccAssetEndpointProfile_update
=== CONT  TestAccAssetEndpointProfile_update
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327192441026174'
Connecting to the VM with IP Address: 52.156.135.133
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAssetEndpointProfile_update (2662.22s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2665.212s

```


ASSET TESTS
===
```
$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAsset_basic'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAsset_basic -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAsset_basic
=== PAUSE TestAccAsset_basic
=== CONT  TestAccAsset_basic
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327153852197306'
Connecting to the VM with IP Address: 20.191.112.206
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAsset_basic (3440.37s)

$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAsset_complete'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAsset_complete -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAsset_complete
=== PAUSE TestAccAsset_complete
=== CONT  TestAccAsset_complete
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327164354762168'
Connecting to the VM with IP Address: 20.191.125.95
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAsset_complete (2495.94s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2518.589s
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        3462.382s

$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAsset_requiresImport'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAsset_requiresImport -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAsset_requiresImport
=== PAUSE TestAccAsset_requiresImport
=== CONT  TestAccAsset_requiresImport
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327173108178724'
Connecting to the VM with IP Address: 52.233.75.241
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAsset_requiresImport (2720.53s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2732.127s

$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAsset_update'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAsset_update -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAsset_update
=== PAUSE TestAccAsset_update
=== CONT  TestAccAsset_update
Running setup steps to SSH into VM and create AIO cluster
Getting public IP address 'acctestpip-250327182101409542'
Connecting to the VM with IP Address: 52.151.30.229
Creating SSH session for stripping carriage return
Stripping carriage return from the setup script
Creating SSH session for enabling execution for setup script
Enabling execution for setup script
Creating SSH session for running setup script
Running setup script
Setup steps to SSH into VM and create AIO cluster completed
--- PASS: TestAccAsset_update (2904.45s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2907.559s
```


TESTS SKIPPED
===
```
$ make acctests SERVICE='deviceregistry'
TF_ACC=1 go test -v ./internal/services/deviceregistry  -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAsset_basic
    device_registry_asset_resource_test.go:726: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAsset_basic (0.00s)
=== RUN   TestAccAsset_complete
    device_registry_asset_resource_test.go:726: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAsset_complete (0.00s)
=== RUN   TestAccAsset_requiresImport
    device_registry_asset_resource_test.go:726: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAsset_requiresImport (0.00s)
=== RUN   TestAccAsset_update
    device_registry_asset_resource_test.go:726: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAsset_update (0.00s)
=== RUN   TestAccAssetEndpointProfile_basic
    device_registry_assetendpointprofile_resource_test.go:685: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAssetEndpointProfile_basic (0.00s)
=== RUN   TestAccAssetEndpointProfile_complete_certificate
    device_registry_assetendpointprofile_resource_test.go:685: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAssetEndpointProfile_complete_certificate (0.00s)
=== RUN   TestAccAssetEndpointProfile_complete_usernamePassword
    device_registry_assetendpointprofile_resource_test.go:685: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAssetEndpointProfile_complete_usernamePassword (0.00s)
=== RUN   TestAccAssetEndpointProfile_complete_anonymous
    device_registry_assetendpointprofile_resource_test.go:685: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAssetEndpointProfile_complete_anonymous (0.00s)
=== RUN   TestAccAssetEndpointProfile_requiresImport
    device_registry_assetendpointprofile_resource_test.go:685: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAssetEndpointProfile_requiresImport (0.00s)
=== RUN   TestAccAssetEndpointProfile_update
    device_registry_assetendpointprofile_resource_test.go:685: Skipping test due to environment variable ARM_ENTRA_APP_OBJECT_ID not set. Required variables: ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_ENTRA_APP_OBJECT_ID
--- SKIP: TestAccAssetEndpointProfile_update (0.00s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2.945s
```

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* new service `DeviceRegistry` - support for the `device_registry` service [GH-00000]
* new resource `azurerm_device_registry_asset` [GH-00000]
* new resource `azurerm_device_registry_asset_endpoint_profile` [GH-00000]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes #0000


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

mryanlo committed:
boilerplate service code

mryanlo committed:
boilerplate service package

mryanlo committed:
AEP resource

mryanlo committed:
fix build errors and lints

mryanlo committed:
lint

mryanlo committed:
cleanup and lint for AEP

mryanlo committed:
update aep + asset resource

mryanlo committed:
vendor dependencies

mryanlo committed:
update codeowners to add device registry required reviewers for deviceregistry tf service

mryanlo committed:
fix build errors in aep and asset resource types

mryanlo committed:
remove untyped service

mryanlo committed:
run `go generate ./internal/provider`

mryanlo committed:
documentation for asset and AEP resources, removing asset and AEP read only fields per terraform requirement

mryanlo committed:
fix aep resource bugs

mryanlo committed:
simplify update logic for aep auth methods

mryanlo committed:
fix bugs in asset resource

mryanlo committed:
documentation example updates

mryanlo committed:
remove attributes fields from documentation

mryanlo committed:
rough draft of tests

mryanlo committed:
Merge branch 'main' of https://github.com/mryanlo/terraform-provider-azurerm into rylo/adr-terraform

mryanlo committed:
add kind.yaml

mryanlo committed:
snake case aep names to fix validation errors

mryanlo committed:
lint

mryanlo committed:
delete aio cluster setup scripts, make asset tests, change aep tests to local cluster only

mryanlo committed:
add comment

mryanlo committed:
add back documentation uri update asset

mryanlo committed:
fix asset test

mryanlo committed:
Merge branch 'main' of https://github.com/mryanlo/terraform-provider-azurerm into rylo/adr-terraform

mryanlo committed:
add github label issue triage

mryanlo committed:
fix lint issues

mryanlo committed:
fix github labeler triage bug by adding the github label method to client registration

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource_test.go:
```diff
@@ -0,0 +1,420 @@
+	r := AssetTestResource{}
+
+	if os.Getenv(ASSET_CUSTOM_LOCATION_NAME) == "" || os.Getenv(ASSET_RESOURCE_GROUP_NAME) == "" {
+		t.Skipf("Skipping test due to missing environment variables %s and/or %s", ASSET_CUSTOM_LOCATION_NAME, ASSET_RESOURCE_GROUP_NAME)
```
For both asset and asset endpoint profile resource tests, we require env vars `ARM_DEVICE_REGISTRY_CUSTOM_LOCATION` and `ARM_DEVICE_REGISTRY_RESOURCE_GROUP` to be set or else the tests will fail (to prevent this we skip the tests).

however, the resource group, custom location, and the AIO cluster must be pre-existing this test run. Investigation and efforts to get the acceptance tests to create and provision an AIO cluster are flakey, take a long time to run (45+ minutes), and we found that the test steps delete the cluster resources before the asset/asset endpoint profile resource tests run (causing the latter ones to fail). awaiting response from Hashicorp and Terraform PMs what the best course of action is here.

stephybun replied:
Hey @mryanlo,

In short it would be acceptable for the tests to run against existing infrastructure for the time being to help move this PR forwards, but the maintainers do need to be able to independently run the acceptance tests for new RPs and resources that are in the provider. I'd like to better understand the procedure required to test these resources so that we can setup the tests to run without a dependence on pre-existing resources.

We have [acceptance tests in the provider for arc enabled clusters](https://github.com/hashicorp/terraform-provider-azurerm/blob/21e34f9573367c7441d67a50844c99cd9f87b49b/internal/services/arckubernetes/arc_kubernetes_cluster_resource_test.go#L30-L52) where we need to provision the prerequisite network resources and a VM that runs the necessary setup scripts for the cluster.

Would you be able to take a look and let me know if there are parts of this setup that you would be able to re-use and what modifications you would need to make for setting up AIO on the cluster specifically? e.g. perhaps the setup script etc.

Once we have a better understanding of what's required we can begin looking at ways to cut down the provisioning time and improving the stability of the test.

mryanlo replied:
Hi @stephybun thanks for reaching out! the exact steps to install AIO onto a cluster are found here: https://review.learn.microsoft.com/en-us/azure/iot-operations/get-started-end-to-end-sample/quickstart-deploy?branch=main

I've seen that link you sent me in the past and have been unable to successfully modify it for AIO's purposes. This is due to the complexity and flakiness of the bash/python script that the VM must run, and our restrictions with the linux distro for installing AIO (you need linux Ubuntu 24.04 with a k3s cluster, not kind cluster to succeed). Additionally there are a bunch of az cli commands that must run and this process is the only available way to install AIO on a cluster, so we need to install az cli and k3s.

If you'd like to see my latest attempt at modifying the scripts to get these to work (the tests will try to sequentially first create a single cluster with AIO installed and then run the acc tests to create resources on only that cluster), here are links: 
- https://github.com/mryanlo/terraform-provider-azurerm/blob/rylo/adr-tf-tests/internal/services/deviceregistry/device_registry_assetendpointprofile_resource_test.go
- https://github.com/mryanlo/terraform-provider-azurerm/tree/rylo/adr-tf-tests/internal/services/deviceregistry/testdata

Thanks, and let me know if you have any more questions about this process.

mryanlo replied:
Note: my code modifies the acceptance tests to run via a managed identity since due to security reasons we are not allowed to make service principals to run terraform anymore.

mryanlo replied:
@stephybun is there a way to ignore the terraform plan check in acceptance tests from running on the VM portions? we don't care about the VM changing since it's not the resource that we want to test. however, one of the issues with my script above is that even if the AIO installation succeeds, the test flags the terraform plan step of the VM as modified after creating.

![image](https://github.com/user-attachments/assets/1c0c9776-206d-4b1f-bdb4-0e8e9cb877ce)

mryanlo replied:
@stephybun I updated the tests in this PR to run completely on their own with all infra setup steps working. In order to do this, I had to use the `PreConfig` step to ssh into the VM and run all the AIO cluster setup script commands, because I noticed the flakiness would happen (the Asset and Asset Endpoint Profile resources would attempt to create even though the setup script didn't finish running via `remote-exec`.). The PreConfig is necessary to block creating all the Assets and Asset Endpoint Profile resources so that it waits for the setup script to finish sequentially before creating the resources. I've tried many things and this is the best solution because having golang sleep or using null_resource did not fix the issue (and would stop the remote-exec steps), and I don't think sleep is a good idea anyways because the setup script can take anywhere between 2000 to 3500 seconds to complete as you see in my screenshots of the results below. Please let me know if this solution is okay. Thank you!

mryanlo replied:
Hi @stephybun  do you have an update on this PR?

mryanlo committed:
update codeowners

mryanlo committed:
change resource logic to poll for create and update functions

mryanlo committed:
fix build

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,743 @@
+					param.Properties.DefaultTopic.Path = pointer.To(config.DefaultTopicPath)
+				}
+				if defaultTopicRetainChanged {
+					// Bug with `go-azure-sdk` library: you can't set retain to null because empty string will cause
```
Linking issue https://github.com/hashicorp/go-azure-sdk/issues/1143 and PR https://github.com/hashicorp/pandora/pull/4633

The go sdk does not support sending null properties. After the PR goes in, this issue can be fixed but might require a code change here

mryanlo committed:
remove comments

mryanlo commented:
Update: figured out a way to get all terraform tests to succeed. Here is the final results:

```
$ make acctests SERVICE='deviceregistry' TESTARGS='-run=TestAccAssetEndpointProfile_basic'
TF_ACC=1 go test -v ./internal/services/deviceregistry -run=TestAccAssetEndpointProfile_basic -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAssetEndpointProfile_basic
=== PAUSE TestAccAssetEndpointProfile_basic
=== CONT  TestAccAssetEndpointProfile_basic
--- PASS: TestAccAssetEndpointProfile_basic (2653.32s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/deviceregistry        2673.792s
```
![image](https://github.com/user-attachments/assets/d094cc9d-f66a-44e9-8945-e42a95c5f1cb)

mryanlo committed:
first attempt at directly sshing from PreConfig in go that works

mryanlo committed:
polish and combine steps to make easier to read

mryanlo committed:
make object id an env variable param, move asset tests to have preconfig block

mryanlo committed:
remove fmt.printf logs

mryanlo committed:
revert getok() logic to get tests working again

mryanlo committed:
Make tests create all AIO Cluster infrastructure (flakey) (#1)

Squash all the commits done to make the asset and AEP resource tests create the VM and AIO cluster in the VM. These steps are flakey because the remote-exec of the AZ CLI commands script in the VM is not awaited and the arc-enabled resources are still attempted to be made which fails the tests.

* start with script + dependency tests on AIO cluster

* update the tests to finally run with completion and success via service principals, prevent resource creation until after resource group created, and vm provisions

* change local resource group name checks to use the terraform resource group object

mryanlo committed:
add comments explaining the steps, add readme to explain overall test process

mryanlo committed:
lint

sreallymatt commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+			ValidateFunc: validation.StringInSlice(assets.PossibleValuesForTopicRetainType(), false),
+			RequiredWith: []string{"default_topic_path"},
+		},
+		"datasets": {
```
Since this is a block, this would usually be singular as multiple blocks could be defined, one per dataset

```suggestion
		"dataset": {
```

sreallymatt commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+						Optional:     true,
+						ValidateFunc: validation.StringInSlice(assets.PossibleValuesForTopicRetainType(), false),
+					},
+					"data_points": {
```
This should be singular

```suggestion
					"data_point": {
```

sreallymatt commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+				},
+			},
+		},
+		"events": {
```
This should be singular
```suggestion
		"event": {
```

sreallymatt reviewed (changes requested):
Hi @mryanlo, I've taken a look over this PR. Just to provide a heads up, as this is a large PR, it will likely take multiple reviews to fully cover everything, I've left some comments inline on my first pass.

I've skipped the test configuration entirely for now, as we'll have to discuss this within the team since it deviates significantly from how we usually do our testing.

mryanlo committed:
revert code owners

mryanlo committed:
Apply suggestions from code review - changes to client.go

Co-authored-by: sreallymatt <106555974+sreallymatt@users.noreply.github.com>

mryanlo committed:
Merge branch 'rylo/adr-terraform' of https://github.com/mryanlo/terraform-provider-azurerm into rylo/adr-terraform

mryanlo commented on internal/services/deviceregistry/device_registry_assetendpointprofile_resource.go:
```diff
@@ -0,0 +1,345 @@
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+		"extended_location_type": {
```
Right now only "CustomLocation" is accepted. So we can put a check there that it is "CustomLocation" or just remove this property entirely

mryanlo replied:
Removed for previous reason

mryanlo commented on internal/services/deviceregistry/device_registry_assetendpointprofile_resource.go:
```diff
@@ -0,0 +1,345 @@
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+		"discovered_asset_endpoint_profile_ref": {
```
no this is not a computed property. The user can essentially put whatever string they want here.

mryanlo commented on internal/services/deviceregistry/device_registry_assetendpointprofile_resource.go:
```diff
@@ -0,0 +1,345 @@
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"authentication_method": {
```
we can make it an `authentication` block. Certificate is the default yes. my question is when do we want to just copy the API pattern vs just take all the properties out of the block and keep them on the root level of the properties object?

mryanlo replied:
made certificate the default and made authentication block.

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"default_topic_path": {
```
no there is no validation to be done here. for most properties, the validation happens at the ARM spec or User RP level. for future reference, when do we want the validation to match that of the ARM spec or UserRP? Or is one validation enough and is replicating it here in terraform redundant?

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+			Optional: true,
+			Elem:     &pluginsdk.Schema{Type: pluginsdk.TypeString},
+		},
+		"discovered_asset_refs": {
```
it is not a computed property. this can be any string value the user desires

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+		"manufacturer": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"manufacturer_uri": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"model": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"product_code": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"hardware_revision": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"software_revision": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"documentation_uri": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"serial_number": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
```
no there is no validation that needs to be done on these properties. These can be whatever strings the user wants with the exception of those that are too long (e.g. 512 chars max i believe). The UserRP already validates this and returns the error to terraform, so we don't need to double check the validation here. Does terraform require validation on all properties?

mryanlo committed:
fix build errors from changing client names to plural. force new true for AEP name

mryanlo committed:
rename aep files to match correct format

mryanlo commented on internal/services/deviceregistry/device_registry_assetendpointprofile_resource.go:
```diff
@@ -0,0 +1,345 @@
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
```
I assume you're talking about the validation like this: 
![image](https://github.com/user-attachments/assets/110850ad-8d13-4a76-857f-a9ad866eff9b)

We can certainly add validation to the name property, but i would like to ask if there is a standard that terraform wants to set for the resource property validation. A lot of our validation is set either at the ARM spec (api reference you mention) level or at the User RP and returns errors to terraform if one of those validation checks fails. As a result, I find this really redundant to keep checking the same validation checks here in terraform especially if the validation changes in the future. But if terraform has some standards to follow, please let me know! thanks

mryanlo commented on website/docs/r/device_registry_asset.html.markdown:
```diff
@@ -0,0 +1,208 @@
+
+* `datasets` - (Optional) Array of datasets that are part of the asset. Each dataset describes the data points that make up the set.
+
+* `default_datasets_configuration` - (Optional) Stringified JSON that contains connector-specific default configuration for all datasets. Each dataset can have its own configuration that overrides the default settings here.
```
we have to use json strings for these `default_` properties. If you look at the documentation, the ARM spec payload indicates it's a stringified json that's expected.

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+		"enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+		},
```
as shown in the [API specs](https://github.com/Azure/azure-rest-api-specs/blob/ba0c086df0ebe03a61579485c1c10de0d17804b2/specification/deviceregistry/resource-manager/Microsoft.DeviceRegistry/stable/2024-11-01/deviceregistry.json#L1156C8-L1159C11), the `enabled` boolean is an optional property. users can set it to be null/nil, or true/false. Because of this support, we can't set a default here

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+		},
+		"external_asset_id": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"display_name": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
+		"description": {
+			Type:     pluginsdk.TypeString,
+			Optional: true,
+		},
```
no there is no validation we could add here and if you look at the ARM specs, there is no validation done on these fields https://github.com/Azure/azure-rest-api-specs/blob/main/specification/deviceregistry/resource-manager/Microsoft.DeviceRegistry/stable/2024-11-01/deviceregistry.json. So we should not add StringIsNotEmpty() validation since they're optional fields too.

mryanlo committed:
make array items singular in name for terraform contract, change names of to/from arm marshallers to match github feedback

mryanlo committed:
use resource group id to parse ID

mryanlo committed:
change terraform "ref"s to "reference"s in asset resource

mryanlo committed:
fix build errors, move aep authentication to nested blocks for simpler logic

mryanlo committed:
change discovered asset endpoint profile ref to references in AEP

mryanlo committed:
fix build error with auth in AEP

mryanlo committed:
change topic to be block type

mryanlo committed:
lint

mryanlo committed:
lint

mryanlo committed:
update the tests for assets

mryanlo committed:
make credential blocks singular

mryanlo commented on website/docs/r/device_registry_asset.html.markdown:
```diff
@@ -0,0 +1,208 @@
+
+* `default_events_configuration` - (Optional) Stringified JSON that contains connector-specific default configuration for all events. Each event can have its own configuration that overrides the default settings here.
+
+* `default_topic` - (Optional) Object that describes the default topic information for the asset.
```
added to the schema

mryanlo committed:
move extended location name to ext location id and remove ext location type since we only support one type

mryanlo committed:
change resource group name to id in tests

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+
+			defaultTopicPathChanged := metadata.ResourceData.HasChange("default_topic_path")
+			defaultTopicRetainChanged := metadata.ResourceData.HasChange("default_topic_retain")
+			if defaultTopicPathChanged || defaultTopicRetainChanged {
```
merged the two properties into one block property

mryanlo committed:
update aep test

mryanlo committed:
fix test checks

mryanlo committed:
fix aep test

mryanlo committed:
fix asset update logic

mryanlo commented on internal/services/deviceregistry/device_registry_asset_resource.go:
```diff
@@ -0,0 +1,742 @@
+		"extended_location_type": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
```
i removed this property since only one value is accepted and will be accepted for the future (CustomLocation)

mryanlo committed:
add custom location validation for extended location id, change credentials back to plural for golang typings in AEP, change the AEP update logic for authentication

mryanlo committed:
remove nested blocks from aep authorization block

mryanlo committed:
fix test aep

mryanlo committed:
fix aep update test

mryanlo commented on website/docs/r/device_registry_asset.html.markdown:
```diff
@@ -0,0 +1,208 @@
+
+* `default_datasets_configuration` - (Optional) Stringified JSON that contains connector-specific default configuration for all datasets. Each dataset can have its own configuration that overrides the default settings here.
+
+* `default_events_configuration` - (Optional) Stringified JSON that contains connector-specific default configuration for all events. Each event can have its own configuration that overrides the default settings here.
```
see my response above

mryanlo commented:
> Hi @mryanlo, I've taken a look over this PR. Just to provide a heads up, as this is a large PR, it will likely take multiple reviews to fully cover everything, I've left some comments inline on my first pass.
> 
> I've skipped the test configuration entirely for now, as we'll have to discuss this within the team since it deviates significantly from how we usually do our testing.

Hi @sreallymatt thank you for the initial review. I've been addressing your feedback. We have a few questions for you and the rest of the team:

1. For adding all of the validation from the ARM spec, we'd like to ask why we must do this here in Terraform? None of our other client-tooling that CRUDs azure resources (sdks, CLIs, etc) do this and they all let the validation happen the RP-level and/or ARM specification. I understand this saves us an API call, but we have changed the validation at various times on the RP and ARM level as requirements came in. We would like to have a single point of validation and maintaining both validations across our own service/ARM and terraform will be tedious. 
2. the reason why this PR is so big is because i had to update the `vendor` folder to include all the go-azure-sdk APIs that the Terraform code would call (generated by running `go mod vendor`). Would it make more sense to separate the vendor folder changes and the service/resource changes into two separate PRs?

sreallymatt commented:
> Hi @sreallymatt thank you for the initial review. I've been addressing your feedback. We have a few questions for you and the rest of the team:
> 
> 1. For adding all of the validation from the ARM spec, we'd like to ask why we must do this here in Terraform? None of our other client-tooling that CRUDs azure resources (sdks, CLIs, etc) do this and they all let the validation happen the RP-level and/or ARM specification. I understand this saves us an API call, but we have changed the validation at various times on the RP and ARM level as requirements came in. We would like to have a single point of validation and maintaining both validations across our own service/ARM and terraform will be tedious.
> 2. the reason why this PR is so big is because i had to update the `vendor` folder to include all the go-azure-sdk APIs that the Terraform code would call (generated by running `go mod vendor`). Would it make more sense to separate the vendor folder changes and the service/resource changes into two separate PRs?

1. From a Terraform provider perspective, it would be a bad user experience to rely on the API to do all of the validation as any validation errors would surface during an apply operation, rather than during a plan. Ideally, once Terraform has planned the changes, and the user has approved them, all changes should be applied without any errors. There are exceptions to this, but generally we aim to validate the user's configuration prior to the apply stage.

2. This is fine, the vendored dependencies are required

mryanlo committed:
lint

mryanlo committed:
more lint changes

mryanlo commented on website/docs/r/device_registry_asset.html.markdown:
````diff
@@ -0,0 +1,208 @@
+Assets can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_device_registry_asset.example C:/Program Files/Git/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/adr-terraform-rg/providers/Microsoft.DeviceRegistry/assets/test-asset
````
this was an issue with git bash using the full computer directory path and not what i entered. fixing in re-generation of the documentation

mryanlo commented on website/docs/r/device_registry_asset_endpoint_profile.html.markdown:
````diff
@@ -0,0 +1,115 @@
+Asset Endpoint Profile can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_device_registry_asset_endpoint_profile.example C:/Program Files/Git/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/adr-terraform-rg/providers/Microsoft.DeviceRegistry/assetendpointprofiles/test-asset-endpoint-profile
````
this was an issue with git bash using the full computer directory path and not what i entered. fixing in re-generation of the documentation

mryanlo commented:
> > Hi @sreallymatt thank you for the initial review. I've been addressing your feedback. We have a few questions for you and the rest of the team:
> > 
> > 1. For adding all of the validation from the ARM spec, we'd like to ask why we must do this here in Terraform? None of our other client-tooling that CRUDs azure resources (sdks, CLIs, etc) do this and they all let the validation happen the RP-level and/or ARM specification. I understand this saves us an API call, but we have changed the validation at various times on the RP and ARM level as requirements came in. We would like to have a single point of validation and maintaining both validations across our own service/ARM and terraform will be tedious.
> > 2. the reason why this PR is so big is because i had to update the `vendor` folder to include all the go-azure-sdk APIs that the Terraform code would call (generated by running `go mod vendor`). Would it make more sense to separate the vendor folder changes and the service/resource changes into two separate PRs?
> 
> Hi @mryanlo, to answer your questions:
> 
> 1. From a Terraform provider perspective, it would be a bad user experience to rely on the API to do all of the validation as any validation errors would surface during an apply operation, rather than during a plan. Ideally, once Terraform has planned the changes, and the user has approved them, all changes should be applied without any errors. There are exceptions to this, but generally we aim to validate the user's configuration prior to the apply stage.
> 2. This is fine, the vendored dependencies are required

Hi @sreallymatt thank you for your reply.

I discussed issue #1 with my team some more and, while we understand what you're saying and would prefer for the validation, we would all like to request if an exception can be made to not add the validation to both Terraform and the Azure cloud service (our API specification in ARM and our service's ResourceProvider/RP). Along with the aforementioned reasons i have provided, some of our validation is very complex and belongs in the cloud service only that cannot be done statically (e.g. validation of unique names of the DataPoints objects in property `Dataset.#.DataPoints.#.Name`, where we must loop through all the data points and validate they have unique names). This kind of logic is too complicated for us to perform on the client/Terraform side. Even if we excluded this complex validation logic from Terraform, it would be a hassle to keep track of the validations we perform on the cloud that are also included in Terraform or not, and maintenance will be difficult moving forward. We recognize that Terraform would prefer to not make the api call and fail during the planning phase, but given all of these requirements on our side, we want to make the call to keep validation at the swagger and RP level and just return the failure back.

mryanlo committed:
documentation asset change

mryanlo committed:
fix paths set by gitbash and change aep documentation

mryanlo committed:
lint

mryanlo commented on website/docs/r/device_registry_asset.html.markdown:
```diff
@@ -0,0 +1,208 @@
+
+* `attributes` - (Optional) A set of key-value pairs that contain custom attributes set by the customer.
+
+* `datasets` - (Optional) Array of datasets that are part of the asset. Each dataset describes the data points that make up the set.
```
done

sreallymatt commented:
@mryanlo, after discussing this with the team we would still like to see the properties validated, at least where possible. This includes regexes for names, max/min values for integers, etc. We recognize this adds a bit of extra maintenance effort, but we do this for almost all other resources to provide a better UX.

The example mentioned (`e.g. validation of unique names of the DataPoints objects in property Dataset.#.DataPoints.#.Name, where we must loop through all the data points and validate they have unique names`), may actually be possible within a `CustomizeDiff` but it would be okay if that validation is left out. (Though, we should still add validation on the `name` field itself, to ensure it matches a regex if the service has naming restrictions)
